Signal/src/util/DateUtil.m

```diff
@@ -187,19 +206,132 @@ + (NSString *)formatDateShort:(NSDate *)date
     return dateTimeString.uppercaseString;
 }
 
-+ (NSString *)formatTimestampAsTimeShort:(uint64_t)timestamp
++ (NSString *)formatTimestampAsTime:(uint64_t)timestamp
 {
-    return [self formatDateAsTimeShort:[NSDate ows_dateWithMillisecondsSince1970:timestamp]];
+    return [self formatDateAsTime:[NSDate ows_dateWithMillisecondsSince1970:timestamp]];
 }
 
-+ (NSString *)formatDateAsTimeShort:(NSDate *)date
++ (NSString *)formatDateAsTime:(NSDate *)date
 {
     OWSAssert(date);
 
     NSString *dateTimeString = [[DateUtil timeFormatter] stringFromDate:date];
     return dateTimeString.uppercaseString;
 }
 
++ (NSDateFormatter *)otherYearMessageFormatter
+{
+    static NSDateFormatter *formatter;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        formatter = [NSDateFormatter new];
+        [formatter setLocale:[NSLocale currentLocale]];
+        [formatter setDateFormat:@"MMM d, yyyy"];
+    });
+    return formatter;
+}
+
++ (NSDateFormatter *)thisYearMessageFormatter
+{
+    static NSDateFormatter *formatter;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        formatter = [NSDateFormatter new];
+        [formatter setLocale:[NSLocale currentLocale]];
+        [formatter setDateFormat:@"MMM d"];
+    });
+    return formatter;
+}
+
++ (NSDateFormatter *)thisWeekMessageFormatter
+{
+    static NSDateFormatter *formatter;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        formatter = [NSDateFormatter new];
+        [formatter setLocale:[NSLocale currentLocale]];
+        [formatter setDateFormat:@"E"];
+    });
+    return formatter;
+}
+
++ (NSString *)formatMessageTimestamp:(uint64_t)timestamp
+{
+    NSDate *date = [NSDate ows_dateWithMillisecondsSince1970:timestamp];
+    uint64_t nowTimestamp = [NSDate ows_millisecondTimeStamp];
+    NSDate *nowDate = [NSDate ows_dateWithMillisecondsSince1970:nowTimestamp];
+
+    NSCalendar *calendar = [NSCalendar currentCalendar];
+
+    NSInteger yearsDiff = [self yearsFromFirstDate:date toSecondDate:nowDate];
+    NSInteger daysDiff = [self daysFromFirstDate:date toSecondDate:nowDate];
+    NSInteger minutesDiff
+        = MAX(0, [[calendar components:NSCalendarUnitMinute fromDate:date toDate:nowDate options:0] minute]);
+    NSInteger hoursDiff
+        = MAX(0, [[calendar components:NSCalendarUnitHour fromDate:date toDate:nowDate options:0] hour]);
+
+    NSString *result;
+    if (yearsDiff > 0) {
```

Comment:
There are some existing methods that I think could clean this up.

e.g.     

if (![self dateIsThisYear:date]) {

} else if (![self dateIsThisWeek:date]) {

} else if (![self isOlderThanToday:date]) {

}

Reply:
Yes, we could do it either way.  I decided to be defensive here to ensure that we treat future dates as "now".  These dates methods are subtle.  I was debating changing the semantics of the existing methods but decided not to - at least, not yet.